### PR TITLE
Check permissions for uploading repository metadata

### PR DESCRIPTION
**Pull request by bklvsky**
Resolves: https://cloudlinux.atlassian.net/browse/ALBS-1211

### CODE REVIEW

File: alws/routers/uploads.py [hunk 1]
```diff
@@ -19,10 +24,17 @@ async def upload_repometada(
     modules: typing.Optional[UploadFile] = None,
     comps: typing.Optional[UploadFile] = None,
     repository: str = Form(...),
+    product_name: str = Form(...),
     session: AsyncSession = Depends(get_db),
+    user: User = Depends(get_current_user),
 ):
     msg = ""
     uploader = MetadataUploader(session, repository)
+    product = get_products(session, product_name=product_name)
```
Comment (javihernandez) [inline alws/routers/uploads.py:33]: Do we really need to add an additional field `product_name` to the endpoint? Can't we use the `product_repositories` table to find the product?

Comment (bklvsky) [reply alws/routers/uploads.py:33]: The repository to update can also be a build repository, for example, and in this case the way to find the product will be different. Maybe just asking to provide the mane of the product is more logical

Comment (javihernandez) [reply alws/routers/uploads.py:33]: Well, my point here is that I'd prefer to guess the product through the repo name provided by the user. The reason?
* I do not have permissions on the `AlmaLinux` product, but I own a personal product called `MyDistro`
* I can call the endpoint indicating an AlmaLinux repo (let's say `baseos_x86_64`) and indicating `MyDistro` product

The current code will allow me to upload metadata to the AlmaLinux repo even though I don't have permissions on that product.

File: alws/routers/uploads.py [hunk 2]
```diff
@@ -18,23 +21,39 @@
...
     user: User = Depends(get_current_user),
 ):
     msg = ""
     uploader = MetadataUploader(session, repository)
-    product = get_products(session, product_name=product_name)
-    if not can_perform(product, user, actions.ReleaseToProduct.name):
-        raise PermissionDenied(
-            "User does not have permissions to upload repository metadata."
-        )
+    repo_objects = await get_repo_object(session, repository)
```
Comment (javihernandez) [inline alws/routers/uploads.py:51]: Looking at uploader code, it currently looks for the repository in pulp directly, see https://github.com/AlmaLinux/albs-web-server/blob/4549f790ae4f8a56ee69417562dd95c9cb74e28a/alws/utils/uploader.py#L276-L278

For this reason, we'd need to do some additional calculation here to figure out the repository in ALBS db, with the provided pulp repo name.

In pulp, production repos naming convention is `distro_name-distro_major_ver-stream-arch`, i.e.: `almalinux-8-baseos-x86_64`.

Naming convention for product repos is: `product_name-distro_name-distro_major_ver-arch-[debug]-dr`, i.e.:
```
soksanichenko-test_product-almalinux-9-x86_64-dr
soksanichenko-test_product-almalinux-9-aarch64-dr
soksanichenko-test_product_2-almalinux-8-aarch64-dr
soksanichenko-test_product_2-almalinux-9-i686-debug-dr
soksanichenko-test_product_2-almalinux-9-i686-dr
soksanichenko-test_product_2-almalinux-9-x86_64-debug-dr
```

Naming convention for build repos is: `distro_name-distro_major_ver-arch-build_id-[debug]-br`, i.e.:
```
AlmaLinux-8-x86_64-76-debug-br
AlmaLinux-9-ppc64le-27-br
AlmaLinux-9-x86_64-28-br
AlmaLinux-8-src-28-br
AlmaLinux-9-x86_64-28-debug-br
```

So, we'd need to first guess the repo (in ALBS) we are going to deal with and then figure out the product it belongs to.

**Review by Korulag — APPROVED** [review]

File: alws/routers/uploads.py [hunk 3]
```diff
@@ -14,15 +22,66 @@
...
+            return (
+                (
+                    await session.execute(
+                        select(Product).filter(Product.team_id == build.team_id)
+                    )
+                )
+                .scalars()
+                .first()
+            )
+    else:
```
Comment (anfimovdm) [inline alws/routers/uploads.py:51]: as for me, this else statement is unnecessary if we will return a value in the if block
```python
if repository.endswith("br"):
    ...
    return product
return (await session.execute(...))
```

File: alws/routers/uploads.py [hunk 4]
```diff
@@ -14,15 +22,66 @@
...
     comps: typing.Optional[UploadFile] = None,
     repository: str = Form(...),
     session: AsyncSession = Depends(get_db),
+    user: User = Depends(get_current_user),
 ):
     msg = ""
     uploader = MetadataUploader(session, repository)
+    repo_product = await get_repo_product(session, repository)
+    if not repo_product:
+        print(f"couldn't find a product or build repository {repository}")
```
Comment (anfimovdm) [inline alws/routers/uploads.py:79]: unnecessary print

File: alws/routers/uploads.py [hunk 5]
```diff
@@ -14,15 +22,66 @@
...
+        return {"error": f"couldn't find a product or build repository {repository}"}
+    if not can_perform(repo_product, user, actions.ReleaseToProduct.name):
```
Comment (anfimovdm) [inline alws/routers/uploads.py:81]: does it work for a product without loading relationships?

**Review by anfimovdm — CHANGES_REQUESTED** [review]

File: alws/routers/uploads.py [hunk 6]
```diff
@@ -14,15 +21,69 @@
 )
 
 
+async def get_repo_product(
```
Comment (javihernandez) [inline alws/routers/uploads.py:24]: I'd say this function can be moved into `alws/crud/products.py`

File: alws/routers/uploads.py (hunk 5 above)
Comment (javihernandez) [reply alws/routers/uploads.py:83]: I'm curious too

**Review by javihernandez — APPROVED** [review]

File: alws/routers/uploads.py (hunk 5 above)
Comment (bklvsky) [reply alws/routers/uploads.py:83]: In the end it didn't, added loading relationships in the previous commit

File: alws/crud/products.py [hunk 7]
```diff
@@ -300,8 +287,7 @@ async def modify_product(
             raise DataNotFoundError(f"User={user_id} doesn't exist")
         if not can_perform(db_product, db_user, actions.ReleaseToProduct.name):
             raise PermissionDenied(
-                'User has no permissions '
-                f'to modify the product "{db_product.name}"'
+                'User has no permissions ' f'to modify the product "{db_product.name}"'
```
Comment (anfimovdm) [inline alws/crud/products.py:290]: please make just one string

File: tests/fixtures/products.py [hunk 8]
```diff
@@ -96,6 +99,33 @@ async def base_product(
...
+                )
+                .options(selectinload(Product.repositories))
+            )
+        )
+        .scalars()
+        .first()
+    )
+    product.repositories.append(repository_for_product)
+    session.add(product)
+    print([repo.name for repo in product.repositories])
```
Comment (anfimovdm) [inline tests/fixtures/products.py:124]: we can remove that print

**Review by anfimovdm — CHANGES_REQUESTED** [review]

File: tests/fixtures/products.py (hunk 8 above)
Comment (bklvsky) [reply tests/fixtures/products.py]: fixed

File: alws/crud/products.py (hunk 7 above)
Comment (bklvsky) [reply alws/crud/products.py]: fixed

**Review by anfimovdm — APPROVED** [review]

**Pull request merged**